### ISSUE TITLE
RAM  Error

### ISSUE THREAD

**Issue by hackedintothemainframe** [jbohnslav/deepethogram#43]
When attempting to train the FlowGenerator using TinyMotionNet and three labelled videos, I get the following error:

`RuntimeError: [enforce fail at ..\c10\core\CPUAllocator.cpp:72] data. DefaultCPUAllocator: not enough memory: you tried to allocate 273715200 bytes. Buy new RAM!`

I'm using a Windows PC with an SSD, 16Gb of DDR4 RAM, and a GTX 1080 with 6 Gb of VRAM. I suspect this may be a CPU issue, as I'm running less than 8 cores (i5 7500), but I wanted to double check that this isn't instead a different issue such as an installation error.

Update: I'm not totally sure about the GPU setting under config.yaml 
`Compute:`
`  batch_size: 32`
`  distributed: false`
`  gpu_id: 0`
`  num_workers: 8`
Is this something that should update automatically, or do I need to physically change the id?

**Comment by jbohnslav** [comment]
Yeah, that's not an error on my end, but PyTorch. Try reducing num_workers to 4 (which should reduce RAM usage)

**Comment by hackedintothemainframe** [comment]
Thanks for the swift response, Jim.

Is the GPU ID of 0 correct?

I think you are indeed correct, the new error log following a change to `num_workers: 4` is:
`...File "C:\Users\hackedintothemainframe\anaconda3\envs\deg\lib\site-packages\torch\utils\data\dataloader.py", line 803, in _get_data`
`  raise RuntimeError('Pin memory thread exited unexpectedly')`
`RuntimeError: Pin memory thread exited unexpectedly`

I also want to double check that my labels should be finalized prior to training the flow generator, correct?

**Comment by hackedintothemainframe** [comment]
To be sure this was not a hardware issue, I installed and ran deepethogram on the same videos, labelled them from scratch, and attempted to train the flowgenerator once again, but on a different computer. This computer runs an SSD, with an 8-core intel Xeon E5-1630, a GTX 1060 (6Gb), 64Gb DDR3 RAM. I tried running flowgenerator twice got the attached logs:

[mbferrorlog.txt](https://github.com/jbohnslav/deepethogram/files/6118851/mbferrorlog.txt)
[mbferrorlog2.txt](https://github.com/jbohnslav/deepethogram/files/6118854/mbferrorlog2.txt)

This seems to be Torch-related once more (and I followed the install instructions closely), but again any help is much appreciated.

**Comment by jbohnslav** [comment]
This is an extremely weird error. One thing to note: you'll never need to re-label videos because of a bug. The labels are simple .CSV files that you can easily move to another computer. 

Can you try training with `num_workers=0`? This will slow down training a lot, so it's not a long-term fix. However, right now the true error is being obscured by the multiprocessing. After that, send me whatever error message you get. 

After you try this and send me the error message, you can try installing the pytorch "nightly" version using instructions here https://pytorch.org/. It looks like this might've been fixed in a recent version of Pytorch: https://github.com/pytorch/pytorch/issues/47445

**Comment by hackedintothemainframe** [comment]
Here is the error log with num_workers=0:
[0workerserrorlog.txt](https://github.com/jbohnslav/deepethogram/files/6143133/0workerserrorlog.txt)

meanwhile I will give pytorch nightly a try!

**Comment by hackedintothemainframe** [comment]
I've installed pytorch nightly, but unfortunately the same error as above is returned. 

One interesting observation is that when monitoring my memory clock while running flowgenerator I usually see a steady memory clock around 9,000 mhz. Right before the crash I get a small spike at 10,010 mhz..

**Comment by hackedintothemainframe** [comment]
I've attempted to do this on my Linux drive as well, and the error I run into with `num_workers: 0` is `RuntimeError: CUDA out of memory. Tried to allocate 7.91 GiB (GPU 0; 7.91 GiB total capacity; 4.09 GiB already allocated; 3.08 GiB free; 4.10 GiB reserved in total by PyTorch)` Has deepethogram been successfully used in cards with 6GiB VRAM?

**Comment by jbohnslav** [comment]
Hi, yes it definitely can be. What resolution are your input videos? In your project_config.yaml, do you have a `resize` field? It might default to 224 x 224. 

You can reduce the batch size in the project config.yaml as well. With 224 x 224 videos and deg_f or deg_m models, you can probably get away with a batch size of 8 or 16.

**Comment by hackedintothemainframe** [comment]
Wow! That was my final problem, the videos were never cropped from 1920x1080. 

**TL;DR for folks running into similar issues in the future: altering the config.yaml settings to `resize: 256` together with attention to CPU cores matching the `num_workers` parameter, and reduction to `batch_size: 4` to accommodate a 6GiB VRAM GPU (a well-loved GTX 1080) allowed the FlowGenerator to run! (however this takes a very, very long time and it's in your best interest to upgrade your hardware!** 

While monitoring the activity during the FlowGenerator, I observed small (in duration, but exceeding 100%) spikes in GPU activity during batches, and large batches (when the setting was > 4) crashed my system even when I limited the max percent usage of my GPU, and I don't know if it's feasible to "smooth" these spikes to improve speed and stability.
I'm using TinyMotionNet, and at the current rate it will take many days to complete. Thankfully this is just a proof of concept pilot run using three ~1 minute long videos with 10 unique behaviors, and better hardware is on its way to the lab. Any further advice on this front is more than welcome. I have also attempted to create a COLAB notebook for free GPU usage, but have been unsuccessful on that front, if anyone has successfully done so, please reach out!

Jim, thanks so much for your time and efforts not only in making this, but making it open-source, and thanks for your help in troubleshooting!

**Comment by jbohnslav** [comment]
> Wow! That was my final problem, the videos were never cropped from 1920x1080.
> 
> **TL;DR for folks running into similar issues in the future: altering the config.yaml settings to `resize: 256` together with attention to CPU cores matching the `num_workers` parameter, and reduction to `batch_size: 4` to accommodate a 6GiB VRAM GPU (a well-loved GTX 1080) allowed the FlowGenerator to run! (however this takes a very, very long time and it's in your best interest to upgrade your hardware!**
> 
> While monitoring the activity during the FlowGenerator, I observed small (in duration, but exceeding 100%) spikes in GPU activity during batches, and large batches (when the setting was > 4) crashed my system even when I limited the max percent usage of my GPU, and I don't know if it's feasible to "smooth" these spikes to improve speed and stability.
> I'm using TinyMotionNet, and at the current rate it will take many days to complete. Thankfully this is just a proof of concept pilot run using three ~1 minute long videos with 10 unique behaviors, and better hardware is on its way to the lab. Any further advice on this front is more than welcome. I have also attempted to create a COLAB notebook for free GPU usage, but have been unsuccessful on that front, if anyone has successfully done so, please reach out!
> 
> Jim, thanks so much for your time and efforts not only in making this, but making it open-source, and thanks for your help in troubleshooting!

Thanks a lot for the detailed response for future reference!

For paper revisions, I'm working on a number of performance improvements and guides, so it should never take many days to train a flow generator. That'll be coming out soon 🤞
